contracts/0.4.24/nos/NodeOperatorsRegistry.sol

```diff
@@ -295,22 +275,41 @@ contract NodeOperatorsRegistry is AragonApp, Versioned {
         emit StakingModuleTypeSet(_type);
     }
 
-    function finalizeUpgrade_v3() external {
-        require(hasInitialized(), "CONTRACT_NOT_INITIALIZED");
-        _checkContractVersion(2);
-        _initialize_v3();
-
-        // clear deprecated total keys count storage
-        Packed64x4.Packed memory summarySigningKeysStats = _loadSummarySigningKeysStats();
-        summarySigningKeysStats.set(SUMMARY_TOTAL_KEYS_COUNT_OFFSET, 0);
-        _saveSummarySigningKeysStats(summarySigningKeysStats);
-    }
-
     function _initialize_v3() internal {
         _setContractVersion(3);
         _updateRewardDistributionState(RewardDistributionState.Distributed);
     }
 
+    function _initialize_v4(
+        uint256 _exitDeadlineThresholdInSeconds,
+        uint256 _reportingWindow
+    ) internal {
+        _setContractVersion(4);
+
+        _setExitDeadlineThreshold(_exitDeadlineThresholdInSeconds, _reportingWindow);
+    }
+
+    /// @notice A function to finalize upgrade to v2 (from v1). Can be called only once.
+    /// For more details see https://github.com/lidofinance/lido-improvement-proposals/blob/develop/LIPS/lip-10.md
+    ///      See historical usage in commit: https://github.com/lidofinance/core/blob/c19480aa3366b26aa6eac17f85a6efae8b9f4f72/contracts/0.4.24/nos/NodeOperatorsRegistry.sol#L230
+    /// function finalizeUpgrade_v2(address _locator, bytes32 _type, uint256 _stuckPenaltyDelay) external
+
+    /// @notice A function to finalize upgrade to v3 (from v2). Can be called only once.
+    /// See historical usage in commit: https://github.com/lidofinance/core/blob/c19480aa3366b26aa6eac17f85a6efae8b9f4f72/contracts/0.4.24/nos/NodeOperatorsRegistry.sol#L298
+    /// function finalizeUpgrade_v3() external
+
+    /// @notice Finalizes upgrade to version 4 by initializing new exit-related parameters.
+    /// @param _exitDeadlineThresholdInSeconds Exit deadline threshold in seconds for validator exits.
```

Comment:
Add link to lip-31 (TBA soon)
    /// For more details see https://github.com/lidofinance/lido-improvement-proposals/blob/develop/LIPS/lip-31.md